openupgrade_scripts/scripts/mail/15.0.1.5/pre-migration.py

```diff
@@ -0,0 +1,67 @@
+from openupgradelib import openupgrade
+
+
+def _copy_columns(env):
+    openupgrade.copy_columns(
+        env.cr,
+        {
+            "mail_activity_type": [
+                ("force_next", None, None),
+                ("res_model_id", None, None),
+            ],
+            "mail_message_res_partner_needaction_rel": [("failure_type", None, None)],
+        },
+    )
+
+
+def _rename_fields(env):
+    openupgrade.rename_fields(
+        env,
+        [
+            (
+                "mail.activity.type",
+                "mail_activity_type",
+                "default_next_type_id",
+                "triggered_next_type_id",
+            ),
+            (
+                "mail.activity.type",
+                "mail_activity_type",
+                "next_type_ids",
+                "suggested_next_type_ids",
+            ),
+            (
+                "mail.activity.type",
+                "mail_activity_type",
+                "default_description",
+                "default_note",
+            ),
+            ("mail.message", "mail_message", "no_auto_thread", "reply_to_force_new"),
+            ("mail.notification", "mail_notification", "mail_id", "mail_mail_id"),
+            ("mail.mail", "mail_mail", "notification", "is_notification"),
+        ],
+    )
+
+
+def _rename_tables(env):
+    openupgrade.rename_tables(
+        env.cr, [("mail_message_res_partner_needaction_rel", "mail_notification")]
+    )
+
+
+def _delete_channel_follower_records(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        DELETE FROM mail_followers
+        WHERE partner_id IS NULL;
+        """,
+    )
+
```

Comment:
Could you please add this ?
```
    # Cleanup potential duplicates to comply with the new constraint mail_channel_partner_partner_unique
    openupgrade.logged_query(
        env.cr,
        """
        DELETE FROM mail_channel_partner mcp
        WHERE
            EXISTS (
                SELECT 1 FROM mail_channel_partner mcp2
                WHERE
                    mcp2.partner_id=mcp.partner_id
                    AND mcp2.channel_id=mcp.channel_id
                HAVING count(*) > 1
            )
            AND NOT EXISTS (
                SELECT 1 FROM ir_model_data imd
                WHERE
                    imd.model ='mail.channel.partner' and res_id = mcp.id
            )
        """,
    )
    
```

Reply:
Assuming that the ones with an ir_model_data entry are created before the other ones, it could be considered to remove the ones with a higher id (using `where exists ... where mcp.id > mcp2.id`). Also, maybe not take `__export__` ir_model_data entries into account.

Reply:
@jcdrubay @StefanRijnhart I have a fix at PR https://github.com/OCA/OpenUpgrade/pull/3400